Refactor session handling

ikari7789 opened the pull request:
Extracted out $_SESSION calls (mostly) to functions and added session handler setting to allow custom session interfaces.

The following mostly removes the direct calls to $_SESSION in the code into functions except for one place where it creates a CookieJar that requires a variable due to it being passed by reference.
It also adds a SessionHandler to the constructor for client so that a user can inject their own custom SessionHandler in cases where they don't want to use the default PHP session handler.

jfritschi commented:
Sorry for the long delay. I was quiet busy IRL.

Can you rework the pull so that I can merge now?

ikari7789 commented:
I've rebased it with master and  updated new code added since this PR was implemented. Is there any thing specific you wanted changed when you said "rework?"

jfritschi commented:
Overall I really like you code.

I think it would be good if you could move more of session code out of the client class. We really should try to have a cleaner structure and not bloat the main class with more functions then needed. This enables better testing and also creates better overview during further development.

Maybe just have a session object that can then be queried? Just a wild thought.

ikari7789 commented:
I actually was thinking that exact same thing as I was looking at this again. Moving the session out of the class though, I wonder if it's not a better idea to implement some 3rd party session handler, such as [Symfony's sessions](http://symfony.com/doc/current/components/http_foundation/sessions.html).

jfritschi commented:
Could be a good idea. I however have never worked with symfony components before.

The question is also if we actually need it. I'm pretty cautious before pulling in new dependencies. Don't want to over complicate our library.... 

Any thoughts on that topic?

dgeo commented:
I'd avoid adding dependencies to phpCAS, which is a dependency of many projects.
The session handling in phpCAS can already break/complicate session handling in dependent app, please do not add anything not strictly needed.
My 2 cents…

thetheodoros commented:
Will this PR ever get merged?
Having custom session interfaces is essential in a microservice architecture context. 
For instance, when using Docker containers you want your container to be ephemeral and have all user/session data handled in a dedicated service.

phy25 commented on source/CAS/Client.php:
```diff
@@ -1074,6 +1086,145 @@ public function getChangeSessionID()
+        if (!isset($_SESSION[static::PHPCAS_SESSION_PREFIX])
+        || !is_array($_SESSION[static::PHPCAS_SESSION_PREFIX])) {
+            $_SESSION[static::PHPCAS_SESSION_PREFIX] = array();
+        }
```
Is this validation needed for every session-related operation? I feel this might be too much...

Just figured out that almost every time CAS_Client is initiated, session will be started. We can probably move this to `CAS_Client.__construct()`?

ikari7789 replied:
What you're suggesting makes sense to me. Is the validation required on every session-related operation? Not sure. If I remember correctly, I think I was just aiming for consistency with existing operation in this PR to keep it easier to review and avoid any big changes in the logic.

We were implementing this into a Laravel project, but basically need to maintain two separate sessions because of how this package is working (Laravel was implementing it's own session management via symfony/session at that time). And implementing this change would've allowed us to integrate it with separate session management methods other than whatever is built into PHP.

ikari7789 commented:
Hi @phy25  I can probably get around to looking at this sometime this week or during the weekend. Gonna take me a bit to re-familiarize myself with a nearly 4-year old PR.

phy25 commented:
Hi @ikari7789, if you want I can take on your work (and still keep your credit), but take your time if you still want to work on it :)

ikari7789 commented:
@phy25 That would be more than appreciated if you can take it on. Sorry, I've just been busy with other projects lately and this one kind of falls lower on the priority scale these days :/

phy25 commented on source/CAS/Client.php:
```diff
@@ -1252,8 +1403,8 @@ public function renewAuthentication()
         // Either way, the user is authenticated by CAS
-        if (isset( $_SESSION['phpCAS']['auth_checked'])) {
-            unset($_SESSION['phpCAS']['auth_checked']);
+        if ($this->hasSessionValue('auth_checked')) {
```
`if` not needed anymore for most `removeSessionValue` calls

phy25 commented on source/CAS/Client.php:
```diff
@@ -1074,6 +1086,145 @@ public function getChangeSessionID()
+        $this->validateSession($key);
+
+        if (isset($_SESSION[static::PHPCAS_SESSION_PREFIX][$key])) {
+            unset($_SESSION[static::PHPCAS_SESSION_PREFIX][$key]);
```
return `true` here so that we know if the key exists before

phy25 commented on source/CAS/Client.php:
```diff
@@ -942,14 +949,12 @@ public function __construct(
+            // TODO remove explicit call to $_SESSION
             $this->_serviceCookieJar = new CAS_CookieJar(
-                $_SESSION['phpCAS']['service_cookies']
+                $_SESSION[static::PHPCAS_SESSION_PREFIX]['service_cookies']
```
I will just keep this as is; the session handler should has been applied here with the reference passing, I guess?

phy25 commented on source/CAS/Client.php:
```diff
@@ -1074,6 +1086,145 @@ public function getChangeSessionID()
+     *
+     * @return void
+     */
+    private function _renameSession($ticket)
```
I am thinking if this should be made protected, not private, so that if anyone needs to customize this along with the session handler, they can do that.

phy25 commented on source/CAS.php:
```diff
@@ -322,20 +322,21 @@ class phpCAS
      */
     public static function client($server_version, $server_hostname,
-        $server_port, $server_uri, $changeSessionID = true
+        $server_port, $server_uri, $changeSessionID = true, SessionHandlerInterface $sessionHandler = null
```
```suggestion
        $server_port, $server_uri, $changeSessionID = true, \SessionHandlerInterface $sessionHandler = null
```

phy25 commented:
#340 is created, closing this.
